### PR TITLE
fix: improve handling of multiple import requests using same file

### PR DESCRIPTION
**Pull request by achou11**
An attempt to potentially (partially) address an issue with making multiple subsequent requests to import the same tileset

Things to note:
- uses a thread pool via [Piscina](https://github.com/piscinajs/piscina)
- switched over the test runner from tap to tape, since tap was running into some race condition issues causing test flakiness
- this ended up being a rather large refactor of tileset imports.

### CODE REVIEW

**Comment by achou11** [comment]
Thinking I can probably remove some code in the worker related to subscriber ids, since each request uses its own worker thread now 🤔

File: src/api.ts [hunk 1]
```diff
@@ -385,23 +388,37 @@ function createApi({
...
         mbTilesDbPath: mbTilesDb.name,
         styleId,
         tilesetId,
       })
 
       return new Promise((res, rej) => {
         // TODO: What else has to be called when this occurs? e.g. terminating the import
-        tilesetImportWorker.addListener('error', (err) => {
+        tilesetImportWorker.on('error', (err) => {
+          activeWorkers.delete(importId)
```
Comment (gmaclennan) [inline src/api.ts:414]: Do we need to terminate the worker here? And similarly below?

**Review by gmaclennan — COMMENTED** [review]
Incomplete review - I've realized I'm too tired to do something thorough. The failing test looks like it needs investigation. Does it pass locally with Node 12?

File: src/api.ts (hunk 1 above)
Comment (achou11) [reply src/api.ts:414]: yeah probably makes sense to for both cases 👍

**Comment by achou11** [comment]
> The failing test looks like it needs investigation. Does it pass locally with Node 12?

Yeah I tried the latest releases of 12, 14, and 16 locally and they all seem to pass. Seems like both 12 and 14 on CI are the flaky ones, as I haven't seen a consistent-enough failure for 16 just yet. Fwiw, looks like the flakiness is present prior to this PR (see CI result in https://github.com/digidem/mapeo-map-server/pull/31), so wondering if some recent versions of node have introduced something 🤔

File: src/api.ts (hunk 1 above)
Comment (achou11) [reply src/api.ts:414]: looking at it more closely, i think we only need to add something for the `error` event case. if the `exit` event is emitted, it's supposed to be the last event emitted by the worker before execution finishes (https://nodejs.org/docs/latest-v12.x/api/worker_threads.html#worker_threads_event_exit)

Comment (achou11) [reply src/api.ts]: See 0d75ffb3f980d7f4bceefa60691974ba6fd7cdd1 for relevant changes

**Comment by achou11** [comment]
Things to do:

- figure out timeout issue happening in CI
- (maybe) throttle progress messages sent by worker (e.g. by time, byte count, row count, etc)

File: package.json [hunk 2]
```diff
@@ -11,12 +11,18 @@
     "dev": "ts-node-dev --files --transpile-only --respawn --no-notify ./server.ts | pino-pretty",
     "build": "npm run clean && tsc -b",
     "clean": "tsc -b --clean",
-    "test": "TS_NODE_FILES=true tap --ts --no-check-coverage --test-ignore='^(dist|scripts)'",
+    "test": "TS_NODE_FILES=true NODE_OPTIONS=\"-r ts-node/register\" c8 ts-node --transpile-only node_modules/tape/bin/tape **/*.test.ts | tap-arc",
```
Comment (gmaclennan) [inline package.json:14]:
```suggestion
    "test": "TS_NODE_FILES=true NODE_OPTIONS=\"-r ts-node/register\" c8 ts-node --transpile-only node_modules/.bin/tape **/*.test.ts | tap-arc",
```

Very slightly more reliable (in case tape changes the location of the bin script in their repo)

**Review by gmaclennan — APPROVED** [review]
So glad this finally works! In the end all the work wasn't needed - it was just a test runner issue thinking the worker exiting was a sigterm error - but I think the code is better now and we cover other potential edge cases.
